zerocopy-derive/src/lib.rs

```diff
@@ -510,105 +512,21 @@ const STRUCT_UNION_ALLOWED_REPR_COMBINATIONS: &[&[StructRepr]] = &[
-        // If the enum could implement `FromBytes`, we can avoid emitting a
-        // match statement. This is faster to compile, and generates code which
-        // performs better.
-        quote!({
-            // Prevent an "unused" warning.
-            let _ = candidate;
-            // SAFETY: If the enum could implement `FromBytes`, then all bit
-            // patterns are valid. Thus, this is a sound implementation.
-            true
-        })
-    } else {
-        quote!(
-            use ::zerocopy::macro_util::core_reexport;
-            // SAFETY:
-            // - The closure is a pointer cast, and `Self` and `[u8;
-            //   size_of::<Self>()]` have the same size, so the returned pointer
-            //   addresses the same bytes as `p` subset of the bytes addressed
-            //   by `slf`
-            // - ..., and so it preserves provenance
-            // - Since we validate that this type is a field-less enum, it
-            //   cannot contain any `UnsafeCell`s. Neither does `[u8; N]`.
-            let discriminant = unsafe { candidate.cast_unsized(|p: *mut Self| p as *mut [core_reexport::primitive::u8; core_reexport::mem::size_of::<Self>()]) };
-            // SAFETY: Since `candidate` has the invariant `Initialized`, we
-            // know that `candidate`'s referent (and thus `discriminant`'s
-            // referent) are fully initialized. Since all of the allowed `repr`s
-            // are types for which all bytes are always initialized, we know
-            // that `discriminant`'s referent has all of its bytes initialized.
-            // Since `[u8; N]`'s validity invariant is just that all of its
-            // bytes are initialized, we know that `discriminant`'s referent is
-            // bit-valid.
-            let discriminant = unsafe { discriminant.assume_valid() };
-            let discriminant = discriminant.read_unaligned();
-
-            false #(|| {
-                let v = Self::#variant_names{};
-                // SAFETY: All of the allowed `repr`s for `Self` guarantee that
-                // `Self`'s discriminant bytes are all initialized. Since we
-                // validate that `Self` has no fields, it has no bytes other
-                // than the discriminant. Thus, it is sound to transmute any
-                // instance of `Self` to `[u8; size_of::<Self>()]`.
-                let d: [core_reexport::primitive::u8; core_reexport::mem::size_of::<Self>()] = unsafe { core_reexport::mem::transmute(v) };
-                // SAFETY: Here we check that the bits of the argument
-                // `candidate` are equal to the bits of a `Self` constructed
-                // using safe code. If this condition passes, then we know that
-                // `candidate` refers to a bit-valid `Self`.
-                discriminant == d
-            })*
-        )
-    };
+    // Generate the equivalent enum in terms of just structs and unions
```

Comment:
This comment feels weird here since it's describing an implementation detail of *how* `derive_is_bit_valid` works.